component/src/lib.rs

```diff
@@ -0,0 +1,127 @@
+// Copyright (c) 2022, Mysten Labs, Inc.
+// SPDX-License-Identifier: Apache-2.0
+
+use anyhow::anyhow;
+use async_trait::async_trait;
+use tokio::sync::mpsc::{channel, Receiver, Sender};
+use tokio::sync::oneshot::{
+    channel as oneshotChannel, Receiver as oneshotReceiver, Sender as oneshotSender,
+};
+
+pub type IrrecoverableError = anyhow::Error;
+type JoinHandle = tokio::task::JoinHandle<()>;
+
+/// A Supervisor is instantiated to supervise a task that should always be running.
+/// A running supervisor will start a component task, and ensure that it is restarted
+/// if it ever stops.
+pub struct Supervisor<M: Manageable> {
+    irrecoverable_signal: Receiver<IrrecoverableError>,
+    cancellation_signal: Option<oneshotSender<()>>,
+    join_handle: Option<JoinHandle>,
+    manageable: M,
+}
+
+/// In order to be Manageable, a user defines the following two functions:
+///
+/// 1. A start function that launches a tokio task, as input it takes:
+/// - an irrecoverable error sender, on which the component sends information to the supervisor about
+/// an irrecoverable event that has occurred
+/// - a cancellation handle, which will be listened to in the task once an irrecoverable message
+/// has been sent, used as an "ack" that the message has been received and so the function can return
+///
+/// 2. A handle_irrecoverable which takes actions on a relaunch due to an irrecoverable error
+/// that happened. It takes the error message that may contain a stack trace and other information
+/// that was sent to the Supervisor via the tx_irrecoverable passed into start.
+#[async_trait]
+pub trait Manageable {
+    // The function that spawns a tokio task
+    async fn start(
+        &self,
+        tx_irrecoverable: Sender<anyhow::Error>,
+        rx_cancellation: oneshotReceiver<()>,
+    ) -> tokio::task::JoinHandle<()>; // Note the task is "silent" (returns nothing)
+
+    // The function for cleanup after the task has encountered an irrecoverable error
+    fn handle_irrecoverable(&self, irrecoverable: IrrecoverableError) -> Result<(), anyhow::Error>;
+}
+
+impl<M: Manageable + Send> Supervisor<M> {
+    /// Creates a new supervisor using a Manageable component.
+    pub fn new(component: M) -> Self {
+        let (_, tr_irrecoverable) = channel(10);
```

Comment:
Do you have a quick intuition on why the channel size if set to 10? Is it a channel we expect to be use only few times?
Maybe silly question: Why is this channel used more than once?